### PR TITLE
Refactor exceptions, code cleanup, and move to `delegate` auth state changes

### PR DESCRIPTION
**Pull request by wiverson**
## What kind of change does this PR introduce?

From a high level, this PR changes the following:

- Dramatically simplifies the exception handling.
- Streamlined error mapping between server errors and the new client exception
- Simplified the persistence plugin functions
- Fixes & hardening for auth state changes
- Adds a bunch of additional tests, including tests for auth state changes and persistence
- Lots of misc doc fixes
- Replaces the console debug write with a function to listen for debug output
- Added info on the changes to the README

Sorry for a lot of changes all at once, hopefully this isn't a giant PITA to review.

I think the most "controversial" thing is tweaking the persistence functions to no longer be async. I decided to remove that as the data involved is tiny, the main thread calling refresh is already running on another thread, and it makes it a lot easier to write the persistence plugin. Could go either way on this one.

I did run all the test suite against the local docker setup. :)

### CODE REVIEW

File: Gotrue/PersistenceListener.cs
Comment (acupofjose) [inline Gotrue/PersistenceListener.cs:1]: Thoughts on something like this instead: https://github.com/supabase-community/supabase-csharp/blob/master/Supabase/Interfaces/ISupabaseSessionHandler.cs ?

**Review by acupofjose — CHANGES_REQUESTED** [review]
HUGE thank you for this one. Also a big help to go through and add some tests.

- I _love_ the change to the exception handler. Much simpler, you were very right about that haha
- The change to async persistence is okay I think. The change to exception handling is a breaking API change anyways, so it's not a huge deal to wrap this in here too.
- I would rather use `EventHandlers` instead of callbacks on AuthState changes.
- There are a few misnamed signatures that need to be fixed!

**Comment by wiverson** [comment]
Awesome. :)

A few As & Qs...

I moved away from EventHandlers because they don't have any way to provide type information, and there are some other language/syntax quirks. I believe that EventHandlers were originally introduced to support some of the UI builder tools back in the day when MS was a lot more interested in that sort of thing. At least in the Unity community they are pretty much considered deprecated nowadays. Switching to straight delegates brings all the usual type safety gains.

RE switching to ISupabaseSessionHandler.cs absolutely, realized I should do that in the shower this morning lol. Keeping them simple delegates and adding this interface will make it lol easy to implement persistence.

Since we are breaking stuff anyways, I think I'm going to change the method names from Persistor to just SaveSession/LoadSession/DestroySession. I think it's easier/clearer plus apparently the spell-checker in my IDE complains about Persistor lol.

RE: SendMagicLink vs SignIn, I found it a bit confusing as the method looked like it was the same as a regular sign in with email. I can change the method name back, but as a user how do I know which is which? This might just be something to add to the doc for the method?

I think I'd like to add one extra method to the API to expose as well, the http://localhost:9999/settings endpoint. I'd like to write a Unity component to make it really easy to get set up with Supabase and this method will make it a lot easier for folks getting started (I can add a widget in the Editor to validate the config is correct).

I'll go ahead and make the changes and send 'em over. I will go ahead and change the method signatures back to SignIn and take a stab at adding my interpretation of what's going on in the flow.

**Comment by wiverson** [comment]
I'm doing a code review one more time and I'm noticing that the SignIn method behavior is inconsistent RE: wiping out the session as a side-effect. Is the general idea that any sign in should wipe out session, or not...?

**Comment by wiverson** [comment]
Ok, I think I got everything. Take a look, LMK if you have any Qs. FWIW I tried flipping the persistence stuff to use an interface, but delegates are not methods but instance variables and so the interface wouldn't work unless I flipped them to methods (adding more indirection complexity) and/or an abstract class (which wound up just being three methods and a constructor anyways). 

The way it works now still works nicely with automatically generating types in the IDE and it's all typesafe, so I think it's fine ¯\_(ツ)_/¯

If you think it looks good I think my next step is testing it out in Unity. I'm hoping to release a demo Unity project that will make it dead simple for someone to just copy and paste their Supbase config into Unity, hit a validate button in the Editor that will ping the new settings, and they'll be off to the races. Hopefully that will be a very short, easy video too. Then I'll add a second video showing the same thing but with Sign in with Apple.

It's a real pity so much of this stuff is such a PITA to test. Oh well.

**Comment by wiverson** [comment]
Ok, now that I'm using the full Supabase I see how the options interact a bit. Minor tweaks to make that easier to work with.

Few other minor things: I think that an abstract class for the persistence class works well. Also adding debug implementations of the various handlers that just write to console to make it easier for folks getting started.

Hoping to do these tonight/tomorrow.

**Comment by acupofjose** [comment]
Okay - I've thought about the delegate changes you proposed all day and I can get behind it!

I'll just have to transfer these paradigm changes to the other repos before doing a release upstream on supabase-csharp as these are going to warrant a semver major release (realtime-csharp has needed some love for a little while anyways and the delegates ought to clean up a lot of the code there!)

Good!

**Comment by wiverson** [comment]
lol you were totally right RE: the persistence interface. Was totally overthinking/overcomplicating it w/delegates, etc. It's now just a simple interface.

**Comment by wiverson** [comment]
Ok, last tweak for now. I moved setting the persistence from the client options to a method. This allows me to set the persistence without relying on the parent Supabase libs to tweak the options there for testing.

**Review by acupofjose — APPROVED** [review]
I've updated the interfaces based on your changes. I like the place you settled on with `PersistenceListener`.... So, LGTM!

**Comment by wiverson** [comment]
Cool, anything you need me to do at this point?

I think the main thing I'd like to do next is beat on it a bit more pointing at a live project to stress a bunch of edge cases. In particular, to make sure the flow works nicely for things that aren't covered by the docker stuff, like the flow for users if auto-confirm is turned off. I also think it would be good to add in some more errors to cover stuff like incorrectly configured nonce values coming back. Mostly I'm hoping those will just be adding more scenarios to the reasons file in the exception. IMHO those can be another PR if you want to close this one, they shouldn't break contracts just add some more hints for when things go sideways.

**Comment by acupofjose** [comment]
Nope - thanks so much. You've been a tremendous help!

**Pull request merged**